### PR TITLE
DataFusion solution [WIP]

### PR DESCRIPTION
**Pull request by Dandandan**
WIP PR to add DataFusion (https://github.com/h2oai/db-benchmark/issues/107) as a solution.

@jangorecki is there any documentation regarding the required output? As this is a Rust solution, it can not easily reuse the prepared code.

### CODE REVIEW

File: datafusion/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    let csv = CsvFile::try_new(&data, options).unwrap();
+    let batch_size = 65536;
+    let memtable = MemTable::load(&csv, batch_size).await?;
+    ctx.register_table("t", Box::new(memtable));
+
+    // "q1"
+    let start = Instant::now();
+    let df = ctx.sql("SELECT id1, SUM(v1) AS v1 FROM t GROUP BY id1")?;
+
+    let _results = df.collect().await?;
```
Comment (jangorecki) [inline datafusion/src/main.rs:39]: I would chain this two lines into single line, variable can be named `ans` for consistency to other scripts. Query for each question need to be run two times. After each query we do extra computation on the `ans` (and measure its timing as well) to ensure that actual query was not lazy by forcing computation on `ans`. Those two timings for each query we needs to be written to csv file with timings, which should be handled by a helper function, like this one: https://github.com/h2oai/db-benchmark/blob/070d3b771c2c57e2834c569fd314f9680a0a2812/_helpers/helpers.py#L8
After both run of a query finished we need to print head-3 and tail-3 of `ans`.

**Review by jangorecki — COMMENTED** [review]
Thank you for submitting this PR. It is really valuable to have a working example for a compiled code in this benchmark, as all currently listed solutions are using REPL interfaces.
I left one comment. I could take it over from here as the most difficult part has been already done. Although it can take me little while because coming week I will be on another project. Anyway, if it is possible for you to translate mentioned `write_log` from python to rust that would be helpful.

**Comment by Dandandan** [comment]
Thanks for the feedback. I replaced the variables by `ans` and combined them in one go. I am not sure if I find enough time for the `write_log` this week.

@jangorecki if you work on it here some tips:

* The solution can be executed with `cargo run --release`. `cargo run` executes it in debug mode, but makes the program very slow.
* The `lto = true` line in `Cargo.toml` speeds up the binary, but slows the build down, so better to remove it while developing.

**Comment by jangorecki** [comment]
Any idea if DataFusion supports queries that are required for advanced groupby questions? q6-q10?

**Comment by Dandandan** [comment]
I added query7 and query 10. Query 10 is ridiculously slow though, that will improve a bit once a PR has been merged, but probably will still be slow after that. The benchmarks showed that we have some more work to do! I also have an open PR that will improve performance on the easier queries, I think DataFusion might already be close for the group by query 1 and 4 to clickhouse / data.table or even CuDF.

The inner join queries should all work too I think (I might add them later, it is easy as I can just reuse the clickhouse queries), there is a known bug for left joins which gives wrong output.

**Comment by matthewmturner** [comment]
@Dandandan anything i can do to help the finalize the work on this?

**Comment by Dandandan** [comment]
> @Dandandan anything i can do to help the finalize the work on this?

Yeah sure, help appreciated!

I think what's missing is:

* Integrating with the standard flow (writing a file in the standardized way)
* Add join queries as well

**Comment by matthewmturner** [comment]
@Dandandan ok! Will check it out. Any additional info you could provide on what the standard flow is?

**Comment by matthewmturner** [comment]
@Dandandan do you have a preference for how i push my updates here?

i started the work here https://github.com/matthewmturner/db-benchmark/tree/datafusion/datafusion as a fork of what you were doing.

ive updated how the tables are created and added the join queries.  still need to review them in more detail / make sure its correct, see if i can add any of the missing group by queries, and i assume we'll want to test the larger datasets as well - but let me know if you have any thoughts.

after the above ill start looking into the flow more.

right now these are the results i get when running the benchmarks:
```
group by
q1 took 56 ms
q2 took 289 ms
q3 took 1305 ms
q4 took 69 ms
q5 took 1158 ms
q7 took 1198 ms
q10 took 24691 ms

join
q1 took 261 ms
q2 took 367 ms
q3 took 334 ms
q4 took 507 ms
q5 took 1936 ms
```

File: datafusion/src/main.rs (hunk 1 above)
Comment (matthewmturner) [reply datafusion/src/main.rs]: hi @jangorecki - im picking up work on this.  do you still need a rust version of `write_log`?

**Comment by Dandandan** [comment]
> @Dandandan do you have a preference for how i push my updates here?
> 
> i started the work here https://github.com/matthewmturner/db-benchmark/tree/datafusion/datafusion as a fork of what you were doing.
> 
> ive updated how the tables are created and added the join queries.  still need to review them in more detail / make sure its correct, see if i can add any of the missing group by queries, and i assume we'll want to test the larger datasets as well - but let me know if you have any thoughts.
> 
> after the above ill start looking into the flow more.
> 
> right now these are the results i get when running the benchmarks:
> ```
> group by
> q1 took 56 ms
> q2 took 289 ms
> q3 took 1305 ms
> q4 took 69 ms
> q5 took 1158 ms
> q7 took 1198 ms
> q10 took 24691 ms
> 
> join
> q1 took 261 ms
> q2 took 367 ms
> q3 took 334 ms
> q4 took 507 ms
> q5 took 1936 ms
> ```

Thank you 💯. Maybe you could open a new PR with the combined changes so we can continue it there?

**Comment by matthewmturner** [comment]
@Dandandan Sure sounds good!

**Comment by Dandandan** [comment]
Follow up PR:

https://github.com/h2oai/db-benchmark/pull/240

**Pull request closed without merging**